### PR TITLE
Converts antimagic from GetComponent to a signal

### PR DESCRIPTION
**Pull request by ninjanomnom**
@XDTM why did you do this to me

### CODE REVIEW

**Comment by XDTM** [comment]
turns out i'm a dumb

File: code/modules/mob/living/living.dm [hunk 1]
```diff
@@ -884,6 +884,20 @@
...
+	if(!magic && !holy)
+		return
+
+	if((magic && has_trait(TRAIT_ANTIMAGIC)) || (holy && has_trait(TRAIT_HOLY)))
+		return src
+
+	var/list/obj/item/item_list = list()
+	item_list |= get_equipped_items(TRUE)
+	item_list |= held_items
+	for(var/obj/O in item_list)
```
Comment (optimumtact) [inline code/modules/mob/living/living.dm]: uhm what the fuck is this? shouldn't you send the signal and just have everything that is antimagic be listening on the mob already?

Comment (optimumtact) [reply code/modules/mob/living/living.dm]: this seems super super wrong

Comment (ninjanomnom) [reply code/modules/mob/living/living.dm]: We haven't gotten around to having memory efficient ways of handling that, signals on anything other than the object itself are difficult to listen to.

The exchange here is less memory cost per antimagic component and less performance cost per equip/unequip in exchange for slightly poorer performance on this signal.

Comment (optimumtact) [reply code/modules/mob/living/living.dm]: what, how do datum components register for signals on their parent obj then that doesn't make any sense

Comment (optimumtact) [reply code/modules/mob/living/living.dm]: like they are getting signals sent to the atom and receiving them

Comment (Cyberboss) [reply code/modules/mob/living/living.dm]: The component is a part of the item, not the mob

File: code/datums/components/anti_magic.dm [hunk 2]
```diff
@@ -4,41 +4,14 @@
...
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	RegisterSignal(COMSIG_ITEM_USER_RECEIVE_MAGIC, .proc/can_protect)
+
 	magic = _magic
 	holy = _holy
 
 /datum/component/anti_magic/proc/can_protect(_magic = TRUE, _holy = FALSE)
-	if(!active)
```
Comment (SpaceManiac) [inline code/datums/components/anti_magic.dm]: either this check should not be dropped or the `active` var should be removed

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/living.dm (hunk 1 above)
Comment (optimumtact) [reply code/modules/mob/living/living.dm]: why can't they register to listen for signals emitted to something that's not their parent

this isn't very flexible

File: code/datums/components/anti_magic.dm (hunk 2 above)
Comment (ninjanomnom) [reply code/datums/components/anti_magic.dm]: I thought this was the component var for if the component should be enabled but it turns out it's just more duplicated functionality.

Comment (XDTM) [reply code/datums/components/anti_magic.dm]: This var is so you can turn off the antimagic without having to add/remote the component itself

Comment (ninjanomnom) [reply code/datums/components/anti_magic.dm]: Components already have a var that does that

Comment (XDTM) [reply code/datums/components/anti_magic.dm]: Oh, right yeah now that it works with signals you can use the default var

**Comment by optimumtact** [comment]
Against my better judgement I'm gonna merge this?

File: code/modules/mob/mob.dm [hunk 3]
```diff
@@ -680,6 +680,15 @@
 			mob_spell_list -= S
 			qdel(S)
 
+/mob/proc/anti_magic_check(magic = TRUE, holy = FALSE)
+	if(!magic && !holy)
+		return
+	var/list/obj/item/item_list = list()
+	item_list |= held_items
+	for(var/obj/O in item_list)
```
Comment (Cyberboss) [inline code/modules/mob/mob.dm]: why are you copying held_items if you don't mutate the list?

File: code/modules/mob/living/living.dm [hunk 4]
```diff
@@ -884,6 +884,20 @@
...
+/mob/living/anti_magic_check(magic = TRUE, holy = FALSE)
+	if(!magic && !holy)
+		return
+
+	if((magic && has_trait(TRAIT_ANTIMAGIC)) || (holy && has_trait(TRAIT_HOLY)))
+		return src
+
+	var/list/obj/item/item_list = list()
+	item_list |= get_equipped_items(TRUE)
+	item_list |= held_items
```
Comment (Cyberboss) [inline code/modules/mob/living/living.dm]: can be simplified as `var/list/obj/item/item_list = held_items | get_equipped_items(TRUE)`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
I can improve on some other stuff here after #38798 is merged so wait on that for now

**Comment by ninjanomnom** [comment]
It's been updated to make use of the signal refactor and other code cleaned up. As a free side benefit I also made it work if applied to mobs directly.

**Pull request merged**